numpy/testing/utils.py

```diff
@@ -1937,3 +1964,150 @@ def __exit__(self, *exc_info):
                 mod.__warningregistry__.clear()
             if mod in self._warnreg_copies:
                 mod.__warningregistry__.update(self._warnreg_copies[mod])
+
+
+class suppressed_warnings(object):
+    """
+    Context manager doing much the same as ``warnings.catch_warnings``.
+    However, it also provides a suppression mechanism, meaning that instead
+    of setting a warning to "ignore", it is possible to add a suppression.
```

Comment:
It's a lot better now. Still not easy to understand, but that's just because it's not trivial to get your head around how warnings really work under the hood:)